stdlib/public/Concurrency/TaskPool.swift

```diff
@@ -0,0 +1,375 @@
+
+// ==== TaskPool --------------------------------------------------------------
+
+/// Starts a new scope that can contain a dynamic number of child tasks.
+/// 
+/// A `TaskPool` is similar to a ``TaskGroup``, however its usage is fairly specialized for
+/// submitting work using child tasks, where the results of those child tasks do not need to be collected.
+/// 
+/// A task pool cannot be iterated over and its child tasks cannot be awaited on explicitly.
+/// 
+/// Task pool tasks are immediately removed from the pool as soon as they complete, 
+/// this is the primary difference from a task pool which stores results (and thus retains the results),
+/// until they are consumed.
+///
+/// Similarly to a ``TaskGroup`` a `TaskPool` awaits all tasks that are submitted to it before returning
+/// from the `withTaskPool` call.
+///
+/// Typical usage of a ``TaskPool`` to handle a stream of incoming requests may look something like this:
+/// 
+/// ```
+/// try await withTaskPool() { pool in
+///     for try await request in self.requests { // may throw on cancellation
+///         pool.addTaskUnlessCancelled {
+///             await handler.handle(request, response: responseWriter) // handler is responsible for writing response
+///         }
+///     }
+/// }
+/// ```
+///
+/// It is typical for servers to want to consume incoming requests as fast as possible, and feed them into the
+/// some `handler` that will perform the actual invocation of user-code and write the response back to the network.
+///
+/// > Note: Note that a `TaskPool` does not guarantee execution order of tasks added using `addTask`,
+/// > they are subject to the same scheduling rules as a `TaskGroup`, and should be used in cases where parallelism
+/// > is desired, rather than a strict one-by-one execution, which can be achieved by not creating additional tasks
+/// > for the processing of the requests.
+/// 
+/// Task Pool Cancellation
+/// =======================
+///
+/// You can cancel a task pool and all of its child tasks
+/// by calling the ``cancelAll()`` method on the task pool,
+/// or by canceling the task in which the pool is running.
+///
+/// If you call `addTask(priority:operation:)` to create a new task in a canceled pool,
+/// that task is immediately canceled after creation.
+/// Alternatively, you can call `asyncUnlessCancelled(priority:operation:)`,
+/// which doesn't create the task if the pool has already been canceled
+/// Choosing between these two functions
+/// lets you control how to react to cancellation within a pool:
+/// some child tasks need to run regardless of cancellation,
+/// but other tasks are better not even being created
+/// when you know they can't produce useful results.
+@available(SwiftStdlib 5.7, *)
+@_unsafeInheritExecutor
+@inlinable
+public func withTaskPool<PoolResult>(
+  returning returnType: PoolResult.Type = PoolResult.self,
+  body: (inout TaskPool) async throws -> PoolResult
+) async rethrows -> PoolResult {
```

Comment:
I don't think rethrows is right here: if the child tasks throw, this throws.

Reply:
Yeah that's wrong. The entire throws handling is not done :) this'll need the same dance as the group, so two with... method versions